plugins/database/quota/src/org/apache/cloudstack/api/response/QuotaResponseBuilderImpl.java

```diff
@@ -0,0 +1,422 @@
+            throw new InvalidParameterValueException("Incorrect effective date for tariff " + effectiveDate + " is less than now " + now);
+        }
+        QuotaTypes quotaConstant = QuotaTypes.listQuotaTypes().get(quotaType);
+        if (quotaConstant == null) {
+            throw new InvalidParameterValueException("Quota type does not exists " + quotaType);
+        }
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        TransactionLegacy.open(TransactionLegacy.USAGE_DB).close();
+        QuotaTariffVO result = null;
+        try {
+            result = new QuotaTariffVO();
+            result.setUsageType(quotaType);
+            result.setUsageName(quotaConstant.getQuotaName());
+            result.setUsageUnit(quotaConstant.getQuotaUnit());
+            result.setUsageDiscriminator(quotaConstant.getDiscriminator());
+            result.setCurrencyValue(quotaCost);
+            result.setEffectiveOn(effectiveDate);
+            result.setUpdatedOn(now);
+            result.setUpdatedBy(cmd.getEntityOwnerId());
+
+            s_logger.debug(String.format("Updating Quota Tariff Plan: New value=%s for resource type=%d effective on date=%s", quotaCost, quotaType, effectiveDate));
+            _quotaTariffDao.addQuotaTariff(result);
+        } catch (Exception pokemon) {
+            s_logger.error("Error in update quota tariff plan: " + pokemon);
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+        return result;
+    }
+
+    @Override
+    public QuotaCreditsResponse addQuotaCredits(Long accountId, Long domainId, Double amount, Long updatedBy) {
+        Date depositDate = new Date();
+        Date adjustedStartDate = _quotaService.computeAdjustedTime(depositDate);
+        QuotaBalanceVO qb = _quotaBalanceDao.findLaterBalanceEntry(accountId, domainId, adjustedStartDate);
+
+        if (qb != null) {
+            throw new InvalidParameterValueException("Incorrect deposit date: " + adjustedStartDate + " there are balance entries after this date");
+        }
+
+        return addQuotaCredits(accountId, domainId, amount, updatedBy, adjustedStartDate);
+    }
+
+    @Override
+    public QuotaCreditsResponse addQuotaCredits(final Long accountId, final Long domainId, final Double amount, final Long updatedBy, final Date despositedOn) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        QuotaCreditsVO result = null;
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            QuotaCreditsVO credits = new QuotaCreditsVO(accountId, domainId, new BigDecimal(amount), updatedBy);
+            s_logger.info("addQuotaCredits: Depositing " + amount + " on adjusted date " + despositedOn);
+            credits.setUpdatedOn(despositedOn);
+            result = _quotaCreditsDao.saveCredits(credits);
+        }
+
+        TransactionLegacy.open(TransactionLegacy.CLOUD_DB).close();
```

Comment:
Why are we accessing the database from a response builder class rather than DAO?